File: model/stability_study.m
function stability_study(varargin)
% STABILITY_STUDY  CFL / Courant-number stability sweep (MATLAB / Octave).
%
% Sweeps gamma^2 around the theoretical 2D CFL boundary (gamma^2 = 0.5),
% runs a short undamped simulation for each value, and classifies each run
% as stable or divergent.  Produces:
%
%   model/outputs/cfl_sweep_envelope.png   log peak-amplitude vs. time
%   model/outputs/cfl_classification.png   stable/divergent bar chart
%   stdout                                 summary table + gamma2_max value
%
% Theory (README §1):
%   The explicit 2D scheme is stable iff  gamma^2 <= 1/2.
%   Crossing the boundary causes exponential divergence.
%
% Name/value options:
%   'plots'  (logical) write PNG plots  [default true; pass false for headless]
%   'outdir' (char)    output directory [default model/outputs/]

    % ---- Sweep configuration ----------------------------------------------
    SWEEP_GAMMA2 = [ ...
        0.10, 0.20, 0.30, 0.40, ...            % well inside stable region
        0.45, 0.48, 0.490, 0.499, ...          % approaching the boundary
        0.500, ...                             % theoretical CFL limit
        0.501, 0.510, 0.52, 0.55, 0.60, 0.70]; % unstable region
    CFL_LIMIT  = 0.5;       % 1/2 — theoretical 2D stability boundary

    C.NX        = 24;       % grid size; small for a fast sweep
    C.NY        = 24;
    C.FS        = 48000;    % Hz
    C.H         = 0.01;     % m, spatial step
    C.SIGMA     = 0.0;      % undamped — cleanest divergence signal
    C.DURATION  = 0.10;     % s  (100 ms per run)
    C.SAMPLE_MS = 0.5;      % amplitude sample interval (ms)
    C.DIV_THRESH = 1e6;     % classify divergent above this peak displacement

    P.plots = true;
    here = fileparts(mfilename('fullpath'));
    P.outdir = fullfile(here, 'outputs');
    P = parse_opts(P, varargin);
    if ~exist(P.outdir, 'dir'); mkdir(P.outdir); end

    fprintf('2D FDTD CFL stability sweep\n');
    fprintf('  Grid     : %dx%d   fs=%g Hz   h=%g m   sigma=%g\n', ...
            C.NX, C.NY, C.FS, C.H, C.SIGMA);
    fprintf('  Duration : %.0f ms per run\n', C.DURATION * 1e3);
    fprintf('  CFL limit: gamma^2 = %g  (gamma = %.4f)\n\n', ...
            CFL_LIMIT, sqrt(CFL_LIMIT));

    header = sprintf('%8s  %8s  %12s  %12s  %12s  div. time', ...
                     'gamma^2', 'gamma', 'status', 'peak@10ms', 'peak@end');
    fprintf('%s\n', header);
    fprintf('%s\n', repmat('-', 1, numel(header)));

    n = numel(SWEEP_GAMMA2);
    results = struct('g2', cell(1, n), 'times', [], 'env', [], 'div_ms', []);

    for i = 1:n
        g2 = SWEEP_GAMMA2(i);
        [times, env, div_ms] = run_one(g2, C);

        idx10 = find(times >= 10.0, 1, 'first');
        if isempty(idx10); idx10 = numel(env); end
        p10  = env(min(idx10, numel(env)));
        pend = env(end);
        if isempty(div_ms)
            status = 'stable';   div_str = '-';
        else
            status = 'DIVERGENT'; div_str = sprintf('%.1f ms', div_ms);
        end

        fprintf('%8.4f  %8.5f  %12s  %12.3e  %12.3e  %s\n', ...
                g2, sqrt(g2), status, p10, pend, div_str);

        results(i).g2 = g2;
        results(i).times = times;
        results(i).env = env;
        results(i).div_ms = div_ms;
    end

    % ---- Recommendation ---------------------------------------------------
    stable_g2 = [results(cellfun(@isempty, {results.div_ms})).g2];
    if isempty(stable_g2); empirical_max = 0.0; else; empirical_max = max(stable_g2); end
    % 10% margin below the empirical limit, rounded to 3 d.p. for clean
    % RTL coefficients.
    gamma2_max = round(empirical_max * 0.90 * 1000) / 1000;

    fprintf('\n%s\n', repmat('-', 1, 60));
    fprintf('Empirical stable maximum : gamma^2 = %.4f\n', empirical_max);
    fprintf('Recommended gamma2_max   : %.3f\n\n', gamma2_max);
    fprintf('Rationale:\n');
    fprintf('  The scheme diverges exponentially for gamma^2 > 0.5 (CFL limit).\n');
    fprintf('  The non-linear term (alpha*u^2, README §2) raises the effective\n');
    fprintf('  local gamma^2 above gamma0^2 on loud transients.  A 10%% margin\n');
    fprintf('  below the empirical limit gives gamma2_max = %.3f, leaving\n', gamma2_max);
    fprintf('  headroom of %.3f for the amplitude-dependent stiffening\n', CFL_LIMIT - gamma2_max);
    fprintf('  before the hard clamp engages.\n\n');

    if ~P.plots; return; end
    % Plotting is a secondary aid; if no graphics toolkit is available
    % (e.g. a bare headless install) skip it rather than failing the study.
    try
        fprintf('Writing plots:\n');
        save_envelope_plot(results, C, P.outdir);
        save_classification_plot(results, gamma2_max, CFL_LIMIT, P.outdir);
        fprintf('Done.\n');
    catch err
        fprintf('  Plotting unavailable (%s) - skipping plots.\n', err.message);
    end
end

% ===========================================================================
% Per-run simulation
% ===========================================================================

function [times_ms, envelope, div_ms] = run_one(gamma2, C)
    % Run one undamped simulation with the given gamma^2.
    % Returns sample times (ms), the peak |u|_inf envelope, and the
    % divergence time in ms (empty if the run stayed stable).
    c = sqrt(gamma2) * C.H * C.FS;
    mesh = Mesh2D(C.NX, C.NY, C.FS, C.H, c, C.SIGMA, 'fixed', false);
    mesh.strike(floor(C.NY/2), floor(C.NX/2), 2.0, 1.0);

    n_total    = round(C.DURATION * C.FS);
    samp_every = max(1, round(C.SAMPLE_MS * 1e-3 * C.FS));

    times_ms = [];
    envelope = [];
    div_ms   = [];

    for n = 1:n_total
        if mod(n - 1, samp_every) == 0
            peak = max(abs(mesh.u(:)));
            bad  = ~isfinite(peak) || peak > C.DIV_THRESH;
            times_ms(end+1) = (n - 1) / C.FS * 1e3; %#ok<AGROW>
            if isfinite(peak)
                envelope(end+1) = min(peak, C.DIV_THRESH * 10); %#ok<AGROW>
            else
                envelope(end+1) = C.DIV_THRESH * 10; %#ok<AGROW>
            end
            if bad
                div_ms = (n - 1) / C.FS * 1e3;
                break;
            end
        end
        mesh.step();
    end
end

% ===========================================================================
% Plots
% ===========================================================================

function save_envelope_plot(results, C, outdir)
    fig = figure('visible', 'off', 'position', [100 100 1100 600]);
    ax = axes(fig); hold(ax, 'on');

    for i = 1:numel(results)
        r = results(i);
        stable = isempty(r.div_ms);
        if stable
            col = [0.2 0.4 0.8]; ls = '-';  lw = 1.2;
        else
            col = [0.85 0.2 0.2]; ls = '--'; lw = 1.5;
        end
        semilogy(ax, r.times, max(r.env, 1e-6), ls, ...
                 'Color', col, 'LineWidth', lw, ...
                 'DisplayName', sprintf('gamma^2=%.3f', r.g2));
    end
    yline_compat(ax, C.DIV_THRESH);

    xlabel(ax, 'Time (ms)');
    ylabel(ax, 'Peak displacement  ||u||_\infty');
    title(ax, {'CFL sweep — peak displacement envelope', ...
               'solid = stable, dashed = divergent'});
    legend(ax, 'Location', 'northwest', 'FontSize', 7, 'NumColumns', 2);
    ylim(ax, [1e-5, C.DIV_THRESH * 50]);

    path = fullfile(outdir, 'cfl_sweep_envelope.png');
    print(fig, path, '-dpng', '-r150');
    close(fig);
    fprintf('  wrote %s\n', path);
end

function save_classification_plot(results, gamma2_max, cfl_limit, outdir)
    g2 = [results.g2];
    fig = figure('visible', 'off', 'position', [100 100 1200 350]);
    ax = axes(fig); hold(ax, 'on');

    for i = 1:numel(results)
        if isempty(results(i).div_ms)
            col = [0.27 0.51 0.71];   % steelblue (stable)
        else
            col = [1.0 0.39 0.28];    % tomato (divergent)
        end
        bar(ax, i, 1, 'FaceColor', col, 'EdgeColor', 'k', 'BarWidth', 0.9);
    end

    cfl_idx = find(g2 >= cfl_limit, 1, 'first');
    if isempty(cfl_idx); cfl_idx = numel(g2) + 1; end
    plot(ax, [cfl_idx-0.5 cfl_idx-0.5], [0 1.05], 'k--', 'LineWidth', 2);

    gmax_idx = find(g2 > gamma2_max, 1, 'first');
    if isempty(gmax_idx); gmax_idx = numel(g2) + 1; end
    plot(ax, [gmax_idx-0.5 gmax_idx-0.5], [0 1.05], ':', ...
         'Color', [0 0.5 0], 'LineWidth', 2);

    set(ax, 'XTick', 1:numel(g2), ...
            'XTickLabel', arrayfun(@(x) sprintf('%.3f', x), g2, ...
                                   'UniformOutput', false), ...
            'YTick', []);
    xtickangle(ax, 45);
    xlabel(ax, 'gamma^2');
    title(ax, sprintf(['CFL classification: stable (blue) / divergent (red)' ...
                       '   —   CFL=%.2f, gamma2\\_max=%.3f'], ...
                      cfl_limit, gamma2_max));
    xlim(ax, [0.5, numel(g2) + 0.5]);
    ylim(ax, [0, 1.1]);

    path = fullfile(outdir, 'cfl_classification.png');
    print(fig, path, '-dpng', '-r150');
    close(fig);
    fprintf('  wrote %s\n', path);
end

% ===========================================================================
% Small helpers
% ===========================================================================

function P = parse_opts(P, args)
    if mod(numel(args), 2) ~= 0
        error('stability_study:args', 'Options must be name/value pairs.');
    end
    for i = 1:2:numel(args)
        name = lower(args{i});
        switch name
            case 'plots';  P.plots  = logical(args{i+1});
            case 'outdir'; P.outdir = args{i+1};
            otherwise
                error('stability_study:args', 'Unknown option "%s".', args{i});
        end
    end
end

function yline_compat(ax, y)
    % yline() is not in older Octave; draw a horizontal reference line.
    xl = xlim(ax);
    if xl(2) <= xl(1); xl = [0 1]; end
    plot(ax, xl, [y y], 'k:', 'LineWidth', 0.8, ...
         'DisplayName', 'div. threshold (1e6)');
end

File: model/stability_study.py
#!/usr/bin/env python3
"""
stability_study.py — CFL / Courant-number stability sweep.

Sweeps gamma^2 around the theoretical 2D CFL boundary (gamma^2 = 0.5),
runs a short undamped simulation for each value, and classifies each run
as stable or divergent.  Produces:

  model/outputs/cfl_sweep_envelope.png   log peak-amplitude vs. time
  model/outputs/cfl_classification.png   stable/divergent bar chart
  stdout                                 summary table + gamma2_max recommendation

Theory (README §1):
  The explicit 2D scheme is stable iff  gamma^2 <= 1/2.
  Crossing the boundary causes exponential divergence.

Usage:
  python model/stability_study.py
  python model/stability_study.py --no-plots
"""

import argparse
import os
import sys

import numpy as np

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from fdtd_ref import Mesh2D

# ---------------------------------------------------------------------------
# Sweep configuration
# ---------------------------------------------------------------------------
SWEEP_GAMMA2 = [
    0.10, 0.20, 0.30, 0.40,           # well inside stable region
    0.45, 0.48, 0.490, 0.499,          # approaching the boundary
    0.500,                              # theoretical CFL limit
    0.501, 0.510, 0.52, 0.55, 0.60, 0.70,  # unstable region
]
CFL_LIMIT = 0.5          # 1/2 — theoretical 2D stability boundary

NX = NY    = 24          # grid size; small for fast sweep
FS         = 48_000      # Hz
H          = 0.01        # m, spatial step
SIGMA      = 0.0         # undamped — cleanest divergence signal
DURATION   = 0.10        # s  (100 ms per run)
SAMPLE_MS  = 0.5         # amplitude sample interval (ms)
DIV_THRESH = 1e6         # classify divergent above this peak displacement


# ---------------------------------------------------------------------------
# Per-run simulation
# ---------------------------------------------------------------------------

def run_one(gamma2):
    """
    Run one undamped simulation with the given gamma^2.

    Returns
    -------
    times_ms  : 1-D float array  — sample times in ms
    envelope  : 1-D float array  — peak |u|_inf at each sample
    div_ms    : float or None    — time of divergence (ms), None if stable
    """
    c = float(np.sqrt(gamma2)) * H * FS
    mesh = Mesh2D(nx=NX, ny=NY, fs=FS, h=H, c=c, sigma=SIGMA,
                  boundary="fixed", check_cfl=False)
    mesh.strike(si=NY // 2, sj=NX // 2, radius=2.0, amp=1.0)

    n_total    = int(DURATION * FS)
    samp_every = max(1, int(SAMPLE_MS * 1e-3 * FS))

    times_ms = []
    envelope = []
    div_ms   = None

    for n in range(n_total):
        if n % samp_every == 0:
            peak = float(np.max(np.abs(mesh.u)))
            bad  = not np.isfinite(peak) or peak > DIV_THRESH
            times_ms.append(n / FS * 1e3)
            envelope.append(min(peak, DIV_THRESH * 10) if np.isfinite(peak) else DIV_THRESH * 10)
            if bad:
                div_ms = n / FS * 1e3
                break
        mesh.step()

    return np.array(times_ms), np.array(envelope), div_ms


# ---------------------------------------------------------------------------
# Plots
# ---------------------------------------------------------------------------

def save_envelope_plot(results, outdir):
    import matplotlib
    matplotlib.use("Agg")
    import matplotlib.pyplot as plt

    fig, ax = plt.subplots(figsize=(11, 6))

    n_stable = sum(1 for *_, d in results if d is None)
    n_div    = len(results) - n_stable
    blues = plt.cm.Blues_r(np.linspace(0.2, 0.7, max(n_stable, 1)))
    reds  = plt.cm.Reds(   np.linspace(0.3, 0.8, max(n_div,    1)))
    bi, ri = 0, 0

    for g2, times, env, div_ms in results:
        label  = f"γ²={g2:.3f}"
        stable = div_ms is None
        color  = blues[bi] if stable else reds[ri]
        ls     = "-" if stable else "--"
        lw     = 1.2 if stable else 1.5
        ax.semilogy(times, np.maximum(env, 1e-6), color=color,
                    ls=ls, lw=lw, label=label)
        if stable:
            bi += 1
        else:
            ri += 1

    ax.axhline(DIV_THRESH, color="k", lw=0.8, ls=":",
               label=f"div. threshold (10⁶)")
    ax.set_xlabel("Time (ms)")
    ax.set_ylabel("Peak displacement  ‖u‖∞")
    ax.set_title(
        "CFL sweep — peak displacement envelope\n"
        "solid = stable, dashed = divergent"
    )
    ax.legend(ncol=2, fontsize=8, loc="upper left")
    ax.set_ylim(1e-5, DIV_THRESH * 50)
    fig.tight_layout()

    path = os.path.join(outdir, "cfl_sweep_envelope.png")
    fig.savefig(path, dpi=150)
    plt.close(fig)
    print(f"  wrote {path}")


def save_classification_plot(results, gamma2_max, outdir):
    import matplotlib
    matplotlib.use("Agg")
    import matplotlib.pyplot as plt
    from matplotlib.patches import Patch
    from matplotlib.lines import Line2D

    g2_vals = [r[0] for r in results]
    colors  = ["steelblue" if r[3] is None else "tomato" for r in results]
    labels  = [f"{g:.3f}" for g in g2_vals]

    fig, ax = plt.subplots(figsize=(12, 3.5))
    ax.bar(range(len(g2_vals)), [1] * len(g2_vals),
           color=colors, edgecolor="k", linewidth=0.5)

    # Mark the CFL boundary between last value < 0.5 and first >= 0.5
    cfl_idx = next((i for i, g in enumerate(g2_vals) if g >= CFL_LIMIT),
                   len(g2_vals))
    ax.axvline(cfl_idx - 0.5, color="k", lw=2, ls="--",
               label=f"CFL limit  γ²=0.5")

    # Mark recommended gamma2_max
    gmax_idx = next((i for i, g in enumerate(g2_vals) if g > gamma2_max),
                    len(g2_vals)) - 0.5
    ax.axvline(gmax_idx, color="darkgreen", lw=2, ls=":",
               label=f"γ²_max={gamma2_max:.3f} (recommended)")

    ax.set_xticks(range(len(g2_vals)))
    ax.set_xticklabels(labels, rotation=45, ha="right", fontsize=9)
    ax.set_xlabel("gamma²")
    ax.set_yticks([])
    ax.set_title("CFL classification: stable (blue) / divergent (red)")
    ax.legend(handles=[
        Patch(color="steelblue", label="stable"),
        Patch(color="tomato",    label="divergent"),
        Line2D([0], [0], color="k",         ls="--", lw=2, label=f"CFL limit γ²=0.5"),
        Line2D([0], [0], color="darkgreen", ls=":",  lw=2, label=f"γ²_max={gamma2_max:.3f}"),
    ], loc="upper right", fontsize=9)
    fig.tight_layout()

    path = os.path.join(outdir, "cfl_classification.png")
    fig.savefig(path, dpi=150)
    plt.close(fig)
    print(f"  wrote {path}")


# ---------------------------------------------------------------------------
# CLI
# ---------------------------------------------------------------------------

def main():
    parser = argparse.ArgumentParser(
        description=__doc__,
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )
    parser.add_argument(
        "--no-plots", action="store_true",
        help="Skip matplotlib output (useful in headless/CI environments)",
    )
    parser.add_argument(
        "--outdir",
        default=os.path.join(os.path.dirname(os.path.abspath(__file__)), "outputs"),
    )
    args = parser.parse_args()
    os.makedirs(args.outdir, exist_ok=True)

    print("2D FDTD CFL stability sweep")
    print(f"  Grid     : {NX}×{NY}   fs={FS} Hz   h={H} m   sigma={SIGMA}")
    print(f"  Duration : {DURATION*1e3:.0f} ms per run")
    print(f"  CFL limit: gamma^2 = {CFL_LIMIT}  (gamma = {CFL_LIMIT**0.5:.4f})")
    print()

    col = f"{'gamma²':>8}  {'gamma':>8}  {'status':>12}  {'peak@10ms':>12}  {'peak@end':>12}  div. time"
    print(col)
    print("─" * len(col))

    results = []
    for g2 in SWEEP_GAMMA2:
        gamma   = g2 ** 0.5
        times, env, div_ms = run_one(g2)

        idx10 = int(np.searchsorted(times, 10.0))
        p10   = env[min(idx10, len(env) - 1)] if len(env) else float("nan")
        pend  = env[-1]                        if len(env) else float("nan")
        status = "stable" if div_ms is None else "DIVERGENT"
        div_str = f"{div_ms:.1f} ms" if div_ms is not None else "—"

        print(f"{g2:>8.4f}  {gamma:>8.5f}  {status:>12}  "
              f"{p10:>12.3e}  {pend:>12.3e}  {div_str}")

        results.append((g2, times, env, div_ms))

    # Recommendation: largest stable gamma^2 known safe, backed off by 10 %
    stable_g2s = [r[0] for r in results if r[3] is None]
    empirical_max = max(stable_g2s) if stable_g2s else 0.0
    # Round down to 3 d.p. to keep a clean value for RTL coefficients
    gamma2_max = float(f"{empirical_max * 0.90:.3f}")

    print()
    print("─" * 60)
    print(f"Empirical stable maximum : gamma^2 = {empirical_max:.4f}")
    print(f"Recommended gamma2_max   : {gamma2_max:.3f}")
    print()
    print("Rationale:")
    print(f"  The scheme diverges exponentially for gamma^2 > 0.5 (CFL limit).")
    print(f"  The non-linear term (alpha*u^2, README §2) raises the effective")
    print(f"  local gamma^2 above gamma0^2 on loud transients.  A 10 % margin")
    print(f"  below the empirical limit gives gamma2_max = {gamma2_max:.3f}, leaving")
    print(f"  headroom of {CFL_LIMIT - gamma2_max:.3f} for the amplitude-dependent stiffening")
    print(f"  before the hard clamp engages.")
    print()

    if args.no_plots:
        return

    try:
        import matplotlib  # noqa: F401
    except ImportError:
        print("matplotlib not available — skipping plots (pip install matplotlib)")
        return

    print("Writing plots:")
    save_envelope_plot(results, args.outdir)
    save_classification_plot(results, gamma2_max, args.outdir)
    print("Done.")


if __name__ == "__main__":
    main()
